### PR TITLE
Text rating

### PR DESCRIPTION
**Pull request by jdelamare**
PR brings in rating feature that takes a vote cast by a student. It checks on the backend to confirm they're allowed to vote then returns a small amount of data to the frontend. The front will take this data, `textId, vote, rating` and update the corresponding text with the new rating.

**How to test**
After checking out the branch, navigate to the text search page and attempt to rate a text that has yet to be read. It should not be possible. Click the text and read it to completion or partially, either works. Afterwards, navigate back to that card on the text search page. You should now be able to rate it. 

Giving it a plus vote should add one to the rating, giving it a minus should subtract one. You can also undo your votes.

**Most affected files**
`text/views/api/text.py`
`web/src/Pages/Text/Search.elm`

### CODE REVIEW

File: user/mixins/models.py [hunk 1]
```diff
@@ -111,4 +112,21 @@ def sections_complete_for(self, text: Text) -> int:
...
+        try:
+            v = TextRating.objects \
+                          .filter(text_id=text.id) \
+                          .first() \
+                          .vote
+            if v == 1:
+                return "up"
+            elif v == 0:
+                raise ValueError
+            else:
```
Comment (bgins) [inline user/mixins/models.py:127]: Might be better to explicitly check for `-1` for down, and make all other values a `ValueError`

File: web/src/Api.elm [hunk 2]
```diff
@@ -236,10 +237,12 @@ credHeader : Cred -> Http.Header
 credHeader (Cred token) =
     Http.header "Authorization" ("Bearer " ++ token)
 
+
 gzipHeader : Http.Header
```
Comment (bgins) [inline web/src/Api.elm:241]: Do we still need this?

File: web/src/Pages/Text/Search.elm [hunk 3]
```diff
@@ -157,6 +162,9 @@ type Msg
     | CloseHint TextSearch.Help.TextHelp
     | PreviousHint
     | NextHint
+      -- rating messages
+    | CastVote Vote Int
+    | Vote (Result (Http.Detailed.Error String) ( Http.Metadata, VoteResponse ))
```
Comment (bgins) [inline web/src/Pages/Text/Search.elm:167]: Let's rename this `GotVote` and rename `CastVote` to `Vote`.

File: web/src/Api.elm (hunk 2 above)
Comment (jdelamare) [reply web/src/Api.elm]: It's gone now

**Review by bgins — APPROVED** [review]
Looks great! Thanks for taking this on. :rocket:

File: web/src/Pages/Text/Search.elm (hunk 3 above)
Comment (jdelamare) [reply web/src/Pages/Text/Search.elm]: changed in commit 5612469

File: user/mixins/models.py (hunk 1 above)
Comment (jdelamare) [reply user/mixins/models.py]: resolved in commit 89d1157

**Pull request merged**